src/main/java/org/jenkinsci/plugins/workflow/support/actions/AnnotatedLogAction.java

```diff
@@ -0,0 +1,327 @@
+                            buf.write(linebuf, prefix.length, len - prefix.length);
+                        }
+                    }
+                    baos.reset();
+                }
+            }
+        } catch (IOException x) {
+            LOGGER.log(Level.WARNING, null, x);
+        }
+        return new AnnotatedLargeText<>(buf, StandardCharsets.UTF_8, !isRunning(node), node);
+    }
+
+    // TODO probably need an API method in LogAction to obtain all log text from a block node and descendants (e.g., a parallel branch)
+
+    /**
+     * Used from <tt>console.jelly</tt> to write annotated log to the given output.
+     */
+    @Restricted(DoNotUse.class) // Jelly
+    public void writeLogTo(long offset, XMLOutput out) throws IOException {
+        getLogText().writeHtmlTo(offset, out.asWriter());
+    }
+
+    /**
+     * Creates a sink to print output from a step.
+     * Will use {@link LogActionImpl} if necessary.
+     * @param node a node which wishes to print output
+     * @param filter an optional log filter (must be {@link Serializable})
+     * @return a stream
+     */
+    @SuppressWarnings("deprecation") // LogActionImpl here for backward compatibility
+    @Restricted(NoExternalUse.class) // for use from DefaultStepContext only
+    public static @Nonnull TaskListener listenerFor(@Nonnull FlowNode node, @CheckForNull ConsoleLogFilter filter) throws IOException, InterruptedException {
+        FlowExecutionOwner owner = node.getExecution().getOwner();
+        if (Util.isOverridden(FlowExecutionOwner.class, owner.getClass(), "getLog", long.class)) {
+            return decorate(owner.getListener(), filter, node);
+        } else { // old WorkflowRun which uses copyLogs
+            return LogActionImpl.stream(node, filter);
+        }
+    }
+
+    /**
+     * Wraps a raw log sink so that each line printed will be annotated as having come from the specified node.
+     * The result is remotable to the extent that the input was.
+     */
+    private static @Nonnull TaskListener decorate(@Nonnull TaskListener raw, @CheckForNull ConsoleLogFilter filter, @Nonnull FlowNode node) {
+        if (node.getAction(AnnotatedLogAction.class) == null) {
+            node.addAction(new AnnotatedLogAction(node));
+        }
+        byte[] prefix = prefix(node);
+        return new DecoratedTaskListener(raw, filter, prefix);
+    }
+
+    /** For testing. */
+    static @Nonnull TaskListener decorate(@Nonnull TaskListener raw, @Nonnull String id) {
+        return new DecoratedTaskListener(raw, null, prefix(id));
+    }
+
+    /**
+     * Unlike {@link FlowNode#isRunning}, handles {@link BlockStartNode}s.
+     */
+    private static boolean isRunning(FlowNode node) {
+        if (node instanceof BlockStartNode) {
+            for (FlowNode head : node.getExecution().getCurrentHeads()) {
+                if (new LinearBlockHoppingScanner().findFirstMatch(head, Predicates.equalTo(node)) != null) {
```

Comment:
Replace with `isActive` from https://github.com/jenkinsci/workflow-api-plugin/pull/45. Looks like this is not the first time I thought to replace the `DepthFirstScanner` with multiple calls to `LinearBlockHoppingScanner`!